subgraphs/enzyme-core/entities/PendleV2Position.ts

```diff
@@ -0,0 +1,95 @@
+import { Address, ethereum } from '@graphprotocol/graph-ts';
+import {
+  ExternalPositionType,
+  PendleV2AllowedMarket,
+  PendleV2Position,
+  PendleV2PositionChange,
+  Vault,
+} from '../generated/schema';
+import { useVault } from './Vault';
+import { getActivityCounter } from './Counter';
+import { logCritical, uniqueEventId } from '@enzymefinance/subgraph-utils';
+import { ensureAsset } from './Asset';
+import { ExternalSdk } from '../generated/contracts/ExternalSdk';
+
+
+export function usePendleV2Position(id: string): PendleV2Position {
+  let position = PendleV2Position.load(id);
+  if (position == null) {
+    logCritical('Failed to load fund {}.', [id]);
+  }
+
+  return position as PendleV2Position;
+}
+
+export function createPendleV2Position(
+  externalPositionAddress: Address,
+  vaultAddress: Address,
+  type: ExternalPositionType,
+): PendleV2Position {
+  let position = new PendleV2Position(externalPositionAddress.toHex());
+  position.vault = useVault(vaultAddress.toHex()).id;
+  position.active = true;
+  position.type = type.id;
+  position.principalTokenHoldings = new Array<string>(0);
+  position.lpTokenHoldings = new Array<string>(0);
+  position.save();
+
+  return position;
+}
+
+export function createPendleV2PositionChange(
+  position: Address,
+  changeType: string,
+  vault: Vault,
+  event: ethereum.Event,
+): PendleV2PositionChange {
+  let change = new PendleV2PositionChange(uniqueEventId(event));
+  change.pendleV2PositionChangeType = changeType;
+  change.externalPosition = position.toHex();
+  change.assets = new Array<string>(0);
+  change.assetAmount = null;
+  change.market = null;
+  change.vault = vault.id;
+  change.timestamp = event.block.timestamp.toI32();
+  change.activityCounter = getActivityCounter();
+  change.activityCategories = ['Vault'];
+  change.activityType = 'Trade';
+  change.save();
+
+  vault.lastAssetUpdate = event.block.timestamp.toI32();
+  vault.save();
+
+  return change;
+}
+
+function pendleV2AllowedMarketId(vault: Vault, marketAddress: Address): string {
+  return vault.id + '/' + marketAddress.toHex()
+}
+
+
+export function ensurePendleV2AllowedMarket(vault: Vault, marketAddress: Address): PendleV2AllowedMarket {
```

Comment:
I would create an additional helper function
```typescript
export function getPendleV2AllowedMarket(vault: Vault, marketAddress: Address): PendleV2AllowedMarket {
  let id = pendleV2AllowedMarketId(vault, marketAddress);

  let market = PendleV2AllowedMarket.load(id);

  if (market == null) {
    logCritical('PendleV2AllowedMarket {} not found', [id]);
  }

  return market as PendleV2AllowedMarket;
}
```
And I would use that function inside also ExternalPositionManager instead of ensure, and also in the `handleMarketRemoved`. I know that we are using ensure pattern in other places of the subgraph, but here I think it only shadows when the market is created. It is only needed in the `MarketAdded` event. If we made a mistake and the market wouldn't be defined then we would create a market with a bad duration which is also incorrect.
With `getPendleV2AllowedMarket` it is easier to understand to intention of the code, and where really the market is created